### PR TITLE
add supervision

### PR DESCRIPTION
**Pull request by alexpantyukhin**

### CODE REVIEW

**Comment by alexpantyukhin** [comment]
@ipinak could you please also review this PR?

**Comment by ipinak** [comment]
I haven't fully reviewed it, but can you use pytest fixtures instead of unittest?
Take a look at this: http://doc.pytest.org/en/latest/fixture.html#fixtures-as-function-arguments

**Comment by alexpantyukhin** [comment]
Actually unittests approach is closer to me, but okay! :)

**Comment by ipinak** [comment]
I suppose you can keep it, but since we use pytest maybe we should stick to it all the way.

**Comment by alexpantyukhin** [comment]
@ipinak Yes, you are right. We need to use one way. I have fixed it!

**Comment by ipinak** [comment]
I'm trying out your branch, but it fails.
This is one of the errors I got
```
def test_add():
        test_pid = PID(address='another_address', id='id')
        lp = LocalProcess(MailBox())

        pr = ProcessRegistry(lambda x: None)
        pr.address = 'address'

>       pr.add('id', lp)

tests/test_processregistry.py:48:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <protoactor.processregistry.ProcessRegistry instance at 0x102bd1518>
id = 'id', aref = <protoactor.process.LocalProcess object at 0x102e17750>

>   ???
E   TypeError: __init__() got an unexpected keyword argument 'ref'
```

File: tests/test_supervision.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+from protoactor.restart_statistics import RestartStatistics
+from protoactor.process import LocalProcess
+from protoactor.mailbox import MailBox
+from protoactor.messages import ResumeMailbox, Restart, Stop
+
+@pytest.fixture(scope='module', )
+def supervisor_data():
+    supervisor = Supervisor()
+    mailbox = MailBox()
+    local_process = LocalProcess(self.mailbox)
```
Comment (ipinak) [inline tests/test_supervision.py]: You need to remove `self` from `self.mailbox` and `self.local_process`. it fails due to that.

**Comment by alexpantyukhin** [comment]
hm.. interesting. I was running tests  and it were passed. Thank you!

File: tests/test_supervision.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    return {
+        'supervisor': supervisor,
+        'mailbox': mailbox,
+        'local_process': local_process,
+        'pid_child': pid_child,
+        'restart_statistic': restart_statistic
+    }
+
+
+    def test_handle_failure_resume_directive(supervisor_data):
```
Comment (ipinak) [inline tests/test_supervision.py]: I think you need to indent out the test functions, because now it seems that they are under `supervisor_data` scope.

**Comment by alexpantyukhin** [comment]
hah! now I see the reason. Identation fail was there.

**Comment by alexpantyukhin** [comment]
fixed and rebased.

**Pull request merged**